### ISSUE TITLE
KOReader crashes on PDF reflow

### ISSUE THREAD

**Issue by dungwinux** [koreader/koreader#10150]
* KOReader version: v2023.01 (android-arm64)
* Device: Nokia X10

#### Issue

KOReader crashes on PDF with reflow enabled. This does not happen on android-arm version.

#### Steps to reproduce

1. Open pdf book
2. Pull up bottom tab, and switch _Reflow_ option to **On**
3. Tap outside to hide the bottom tab (It doesn't crash immediately)
4. Tap/Swipe to next page

##### `crash.log`

[crash.log](https://github.com/koreader/koreader/files/10788040/crash.log)

**Comment by NiLuJe** [comment]
Will quite likely magically go away once we start using an actually sensible toolchain for Android ;).

**Comment by neo200321** [comment]
It happened exactly the same way on bigme inknote color. Look forward to solve.

- KOReader version: v2023.01 (android-arm64)
- Device: Bigme inknote color

[crash.log](https://github.com/koreader/koreader/files/10928385/crash.log)
[reports.txt](https://github.com/koreader/koreader/files/10928387/reports.txt)

**Comment by colinkarr** [comment]
This also occurred for me. However, the crash occurred immediately after tapping away from the bottom menu.

- KOReader Version: v2023.01
- Device: Pixel 6a
- crash log was blank

**Comment by amitbha** [comment]
The same problem on Xiaomi cellphone.

    KOReader version: v2023.03 (android-arm64)
    Device: Xiaomi grus

KR crashes occasionally when touch the screen.
[crash-log.txt](https://github.com/koreader/koreader/files/11133815/KR-crash.txt)

**Comment by petitengineer** [comment]
I've had the same issue on a Galaxy S10.

The crash is consistent when I tap for the next page.

[crashlog.txt](https://github.com/koreader/koreader/files/11174584/crashlog.txt)

**Comment by seth-aung** [comment]
This also occurred for me. However, the crash occurred immediately after tapping away from the bottom menu.

> KOReader Version: v2023.03 (arm64)

> Device: Infinix Hot 11s

> crash log was blank

**Comment by beelzebielsk** [comment]
Reflow didn't used to crash koreader on Android. I don't remember when specifically it stopped working.

**Comment by seth-aung** [comment]
It still crushed when I reflow. Sometime, immediately after tapping away from the bottom menu and sometime, tap to next page.
- KOReader Version: v2023.04 (arm64)
- Device: Infinix Hot 11s
[Crash Log 1.txt](https://github.com/koreader/koreader/files/11395498/Crash.Log.1.txt)
[Crash Log 2.txt](https://github.com/koreader/koreader/files/11395499/Crash.Log.2.txt)
[Crash Log 3.txt](https://github.com/koreader/koreader/files/11395500/Crash.Log.3.txt)
[Crash Log 4.txt](https://github.com/koreader/koreader/files/11395501/Crash.Log.4.txt)
[Crash Log 5.txt](https://github.com/koreader/koreader/files/11395502/Crash.Log.5.txt)

**Comment by NiLuJe** [comment]
> Reflow didn't used to crash koreader on Android. I don't remember when specifically it stopped working.

This one is specific to the arm64 build because old GCC is awful.

**Comment by hocandres** [comment]
This also occurred for me. The crash occurred immediately after tapping away from the bottom menu.

    - KOReader Version: v2023.05
    - Device: Pixel 7 Pro
    - Android version: 13
    - Crash log was blank

**Comment by amitbha** [comment]
I will recommend an excellent APP for reading PDF: WPS Office, which supports:
- text reflow
- annotations
- export highlights

**Comment by ryanwwest** [comment]
FWIW, I've been running into the same described PDF reflow problem on the MacOS build and on Android for months, including on a build from yesterday, and that macos build seems to be x86, not arm64..? : https://github.com/koreader/koreader/actions/runs/5451414944/jobs/9917640957#step:5:21635. Maybe I'm misunderstanding it and it does apply to the arm64 solution updating to clang.

**Comment by Frenzie** [comment]
I think the described problem is for any reflow at all. There's another problem with reflow+Chinese (=character) selection.

**Comment by LucaReggiannini** [comment]
Same for me:

- KOReader v2023.06.01
- POCO F3
- Lineage 13 for Alioth v 20230804

crash log was blank

**Comment by NiLuJe** [comment]
To interested parties:

As I mentioned in one of the various duplicates of this issue, it looks (I haven't checked, no time nor interest) like it crashes on hinting (i.e., it reflows the current page just fine, and will only crash when paging).

It might be a bionic-specific quirk in the layout of the pthreads struct, as I remember the situation being pretty funky already on Linux.

**Comment by NiLuJe** [comment]
Eeh, or not.

https://android.googlesource.com/platform/bionic/+/master/libc/include/bits/pthread_types.h

Haven't double-checked, but napkin maths says it's a large as on x64 Linux (once you account for alignment).

Oh, well.

**Comment by NiLuJe** [comment]
@Frenzie: Reminder for your pre-release testing?

**Comment by NiLuJe** [comment]
Not necessarily, that was just me wondering why you'd re-opened this one ;).
